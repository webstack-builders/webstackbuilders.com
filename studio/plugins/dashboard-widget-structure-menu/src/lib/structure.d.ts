export declare function isSubscribable(thing: any): any;
export declare function isStructure(structure: any): any;
export declare function serializeStructure(item: any, context: any, resolverArgs?: never[]): any;
export declare function getDefaultStructure(): any;
export declare function loadStructure(): import("rxjs").Observable<unknown>;
//# sourceMappingURL=structure.d.ts.map