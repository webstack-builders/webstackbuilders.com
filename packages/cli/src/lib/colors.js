import chalk from 'chalk'

/**
 * To keep a consistent color/style palette between cli packages, such as
 * @redwood/cli and @redwood/create-redwood-app, please keep them compatible
 * with one and another. We'll might split up and refactor these into a
 * separate package when there is a strong motivation behind it.
 *
 * Current files:
 *
 * - packages/cli/src/lib/colors.js (this file)
 * - packages/create-redwood-app/src/create-redwood-app.js
 */
export default {
  error: chalk.bold.red,
  warning: chalk.keyword('orange'),
  green: chalk.green,
  info: chalk.grey,
  bold: chalk.bold,
  underline: chalk.underline,
}
