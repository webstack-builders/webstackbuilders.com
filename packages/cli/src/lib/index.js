import fs from 'fs'
import https from 'https'
import path from 'path'

import * as babel from '@babel/core'
import { getDMMF } from '@prisma/sdk'
import camelcase from 'camelcase'
import decamelize from 'decamelize'
import execa from 'execa'
import Listr from 'listr'
import VerboseRenderer from 'listr-verbose-renderer'
import lodash from 'lodash/string'
import { paramCase } from 'param-case'
import pascalcase from 'pascalcase'
import pluralize from 'pluralize'
import { format } from 'prettier'

import {
  getPaths as getRedwoodPaths,
  getConfig as getRedwoodConfig,
} from '@redwoodjs/internal'

import c from './colors'

/**
 * Used to memoize results from `getSchema` so we don't have to go through
 * the work of open the file and parsing it from scratch each time getSchema()
 * is called with the same model name.
 */
const schemaMemo = {}

export const asyncForEach = async (array, callback) => {
  for (let index = 0; index < array.length; index++) {
    await callback(array[index], index, array)
  }
}

/**
 * Returns the database schema for the given `name` database table parsed from
 * the schema.prisma of the target application. If no `name` is given then the
 * entire schema is returned.
 */
export const getSchema = async (name) => {
  if (name) {
    if (!schemaMemo[name]) {
      const schema = await getSchemaDefinitions()
      const model = schema.datamodel.models.find((model) => {
        return model.name === name
      })

      if (model) {
        // look for any fields that are enums and attach the possible enum values
        // so we can put them in generated test files
        model.fields.forEach((field) => {
          const fieldEnum = schema.datamodel.enums.find((e) => {
            return field.type === e.name
          })
          if (fieldEnum) {
            field.enumValues = fieldEnum.values
          }
        })

        // memoize based on the model name
        schemaMemo[name] = model
      } else {
        throw new Error(
          `No schema definition found for \`${name}\` in schema.prisma file`
        )
      }
    }
    return schemaMemo[name]
  } else {
    return (await getSchemaDefinitions()).metadata.datamodel
  }
}

/**
 * Returns the enum defined with the given `name` parsed from
 * the schema.prisma of the target applicaiton. If no `name` is given then the
 * all enum definitions are returned
 */
export const getEnum = async (name) => {
  const schema = await getSchemaDefinitions()

  if (name) {
    const model = schema.datamodel.enums.find((model) => {
      return model.name === name
    })

    if (model) {
      return model
    } else {
      throw new Error(
        `No enum schema definition found for \`${name}\` in schema.prisma file`
      )
    }
  }

  return schema.metadata.datamodel.enums
}

/*
 * Returns the DMMF defined by `prisma` resolving the relevant `schema.prisma` path.
 */
export const getSchemaDefinitions = async () => {
  const schemaPath = path.join(getPaths().api.db, 'schema.prisma')
  const metadata = await getDMMF({
    datamodel: readFile(schemaPath.toString()),
  })

  return metadata
}

/**
 * Returns variants of the passed `name` for usage in templates. If the given
 * name was "fooBar" then these would be:

 * pascalName: FooBar
 * singularPascalName: FooBar
 * pluralPascalName: FooBars
 * singularCamelName: fooBar
 * pluralCamelName: fooBars
 * singularParamName: foo-bar
 * pluralParamName: foo-bars
 * singularConstantName: FOO_BAR
 * pluralConstantName: FOO_BARS
*/
export const nameVariants = (name) => {
  const normalizedName = pascalcase(paramCase(pluralize.singular(name)))

  return {
    pascalName: pascalcase(name),
    camelName: camelcase(name),
    singularPascalName: normalizedName,
    pluralPascalName: pluralize(normalizedName),
    singularCamelName: camelcase(normalizedName),
    pluralCamelName: camelcase(pluralize(normalizedName)),
    singularParamName: paramCase(normalizedName),
    pluralParamName: paramCase(pluralize(normalizedName)),
    singularConstantName: decamelize(normalizedName).toUpperCase(),
    pluralConstantName: decamelize(pluralize(normalizedName)).toUpperCase(),
  }
}

export const templateRoot = path.resolve(__dirname, '../commands/generate')

export const generateTemplate = (templateFilename, { name, root, ...rest }) => {
  const templatePath = path.join(root || templateRoot, templateFilename)
  const template = lodash.template(readFile(templatePath).toString())

  const renderedTemplate = template({
    name,
    ...nameVariants(name),
    ...rest,
  })

  return prettify(templateFilename, renderedTemplate)
}

export const prettify = (templateFilename, renderedTemplate) => {
  // We format .js and .css templates, we need to tell prettier which parser
  // we're using.
  // https://prettier.io/docs/en/options.html#parser
  const parser = {
    '.css': 'css',
    '.js': 'babel',
    '.ts': 'babel-ts',
  }[path.extname(templateFilename.replace('.template', ''))]

  if (typeof parser === 'undefined') {
    return renderedTemplate
  }

  return format(renderedTemplate, {
    ...prettierOptions(),
    parser,
  })
}

export const readFile = (target) => fs.readFileSync(target)

const SUPPORTED_EXTENSIONS = ['.js', '.ts', '.tsx']

export const deleteFile = (file) => {
  const extension = path.extname(file)
  if (SUPPORTED_EXTENSIONS.includes(extension)) {
    const baseFile = getBaseFile(file)
    SUPPORTED_EXTENSIONS.forEach((ext) => {
      const f = baseFile + ext
      if (fs.existsSync(f)) {
        fs.unlinkSync(f)
      }
    })
  } else {
    fs.unlinkSync(file)
  }
}

const getBaseFile = (file) => file.replace(/\.\w*$/, '')

const existsAnyExtensionSync = (file) => {
  const extension = path.extname(file)
  if (SUPPORTED_EXTENSIONS.includes(extension)) {
    const baseFile = getBaseFile(file)
    return SUPPORTED_EXTENSIONS.some((ext) => fs.existsSync(baseFile + ext))
  }

  return fs.existsSync(file)
}

export const writeFile = (
  target,
  contents,
  { overwriteExisting = false } = {},
  task = {}
) => {
  const { base } = getPaths()
  task.title = `Writing \`./${path.relative(base, target)}\``
  if (!overwriteExisting && fs.existsSync(target)) {
    throw new Error(`${target} already exists.`)
  }

  const filename = path.basename(target)
  const targetDir = target.replace(filename, '')
  fs.mkdirSync(targetDir, { recursive: true })
  fs.writeFileSync(target, contents)
  task.title = `Successfully wrote file \`./${path.relative(base, target)}\``
}

export const saveRemoteFileToDisk = (
  url,
  localPath,
  { overwriteExisting = false } = {}
) => {
  if (!overwriteExisting && fs.existsSync(localPath)) {
    throw new Error(`${localPath} already exists.`)
  }

  const downloadPromise = new Promise((resolve, reject) =>
    https.get(url, (response) => {
      if (response.statusCode === 200) {
        response.pipe(fs.createWriteStream(localPath))
        resolve()
      } else {
        reject(
          new Error(`${url} responded with status code ${response.statusCode}`)
        )
      }
    })
  )

  return downloadPromise
}

export const getInstalledRedwoodVersion = () => {
  try {
    const packageJson = require('../../package.json')
    return packageJson.version
  } catch (e) {
    console.error(c.error('Could not find installed redwood version'))
    process.exit(1)
  }
}

export const bytes = (contents) => Buffer.byteLength(contents, 'utf8')

/**
 * This wraps the core version of getPaths into something that catches the exception
 * and displays a helpful error message.
 */
export const getPaths = () => {
  try {
    return getRedwoodPaths()
  } catch (e) {
    console.error(c.error(e.message))
    process.exit(1)
  }
}

export const getConfig = () => {
  try {
    return getRedwoodConfig()
  } catch (e) {
    console.error(c.error(e.message))
    process.exit(1)
  }
}

/**
 * This returns the config present in `prettier.config.js` of a Redwood project.
 */
export const prettierOptions = () => {
  try {
    return require(path.join(getPaths().base, 'prettier.config.js'))
  } catch (e) {
    return undefined
  }
}

// TODO: Move this into `generateTemplate` when all templates have TS support
/*
 * Convert a generated TS template file into JS.
 */
export const transformTSToJS = (filename, content) => {
  const { code } = babel.transform(content, {
    filename,
    // If you ran `yarn rw generate` in `./web` transformSync would import the `.babelrc.js` file,
    // in `./web`? despite us setting `configFile: false`.
    cwd: process.env.NODE_ENV === 'test' ? undefined : getPaths().base,
    configFile: false,
    plugins: [
      [
        '@babel/plugin-transform-typescript',
        {
          isTSX: true,
          allExtensions: true,
        },
      ],
    ],
    retainLines: true,
  })

  return prettify(filename.replace(/\.ts$/, '.js'), code)
}

/**
 * Creates a list of tasks that write files to the disk.
 *
 * @param files - {[filepath]: contents}
 */
export const writeFilesTask = (files, options) => {
  const { base } = getPaths()
  return new Listr(
    Object.keys(files).map((file) => {
      const contents = files[file]
      return {
        title: `...waiting to write file \`./${path.relative(base, file)}\`...`,
        task: (ctx, task) => writeFile(file, contents, options, task),
      }
    })
  )
}

/**
 * Creates a list of tasks that delete files from the disk.
 *
 * @param files - {[filepath]: contents}
 */
export const deleteFilesTask = (files) => {
  const { base } = getPaths()

  return new Listr([
    ...Object.keys(files).map((file) => {
      return {
        title: `Destroying \`./${path.relative(base, getBaseFile(file))}\`...`,
        skip: () => !existsAnyExtensionSync(file) && `File doesn't exist`,
        task: () => deleteFile(file),
      }
    }),
    {
      title: 'Cleaning up empty directories...',
      task: () => cleanupEmptyDirsTask(files),
    },
  ])
}

/**
 * @param files - {[filepath]: contents}
 * Deletes any empty directrories that are more than three levels deep below the base directory
 * i.e. any directory below /web/src/components
 */
export const cleanupEmptyDirsTask = (files) => {
  const { base } = getPaths()
  const endDirs = Object.keys(files).map((file) => path.dirname(file))
  const uniqueEndDirs = [...new Set(endDirs)]
  // get the additional path directories not at the end of the path
  const pathDirs = []
  uniqueEndDirs.forEach((dir) => {
    const relDir = path.relative(base, dir)
    const splitDir = relDir.split(path.sep)
    splitDir.pop()
    while (splitDir.length > 3) {
      const subDir = path.join(base, splitDir.join('/'))
      pathDirs.push(subDir)
      splitDir.pop()
    }
  })
  const uniqueDirs = uniqueEndDirs.concat([...new Set(pathDirs)])

  return new Listr(
    uniqueDirs.map((dir) => {
      return {
        title: `Removing empty \`./${path.relative(base, dir)}\`...`,
        task: () => fs.rmdirSync(dir),
        skip: () => {
          if (!fs.existsSync(dir)) {
            return `Doesn't exist`
          }
          if (fs.readdirSync(dir).length > 0) {
            return 'Not empty'
          }
          return false
        },
      }
    })
  )
}

const wrapWithSet = (routesContent, layout, routes, newLineAndIndent) => {
  const [_, indentOne, indentTwo] = routesContent.match(
    /([ \t]*)<Router.*?>[^<]*[\r\n]+([ \t]+)/
  ) || ['', 0, 2]
  const oneLevelIndent = indentTwo.slice(0, indentTwo.length - indentOne.length)
  const newRoutesWithExtraIndent = routes.map((route) => oneLevelIndent + route)
  return [`<Set wrap={${layout}}>`, ...newRoutesWithExtraIndent, `</Set>`].join(
    newLineAndIndent
  )
}

/**
 * Update the project's routes file.
 */
export const addRoutesToRouterTask = (routes, layout) => {
  const redwoodPaths = getPaths()
  const routesContent = readFile(redwoodPaths.web.routes).toString()
  const newRoutes = routes.filter((route) => !routesContent.match(route))

  if (newRoutes.length) {
    const [routerStart, newLineAndIndent] =
      routesContent.match(/\s*<Router.*?>(\s*)/)
    const routesBatch = layout
      ? wrapWithSet(routesContent, layout, newRoutes, newLineAndIndent)
      : newRoutes.join(newLineAndIndent)
    const newRoutesContent = routesContent.replace(
      routerStart,
      `${routerStart + routesBatch + newLineAndIndent}`
    )
    writeFile(redwoodPaths.web.routes, newRoutesContent, {
      overwriteExisting: true,
    })
  }
}

export const addScaffoldImport = () => {
  const appJsPath = getPaths().web.app
  let appJsContents = readFile(appJsPath).toString()

  if (appJsContents.match('./scaffold.css')) {
    return 'Skipping scaffold style include'
  }

  appJsContents = appJsContents.replace(
    "import Routes from 'src/Routes'\n",
    "import Routes from 'src/Routes'\n\nimport './scaffold.css'"
  )
  writeFile(appJsPath, appJsContents, { overwriteExisting: true })

  return 'Added scaffold import to App.{js,tsx}'
}

const removeEmtpySet = (routesContent, layout) => {
  const setWithLayoutReg = new RegExp(
    `\\s*<Set[^>]*wrap={${layout}}[^<]*>([^<]*)<\/Set>`
  )
  const [matchedSet, childContent] = routesContent.match(setWithLayoutReg) || []
  if (!matchedSet) {
    return routesContent
  }

  const child = childContent.replace(/\s/g, '')
  if (child.length > 0) {
    return routesContent
  }
  return routesContent.replace(setWithLayoutReg, '')
}

/**
 * Remove named routes from the project's routes file.
 *
 * @param {string[]} routes - Route names
 */
export const removeRoutesFromRouterTask = (routes, layout) => {
  const redwoodPaths = getPaths()
  const routesContent = readFile(redwoodPaths.web.routes).toString()
  const newRoutesContent = routes.reduce((content, route) => {
    const matchRouteByName = new RegExp(`\\s*<Route[^>]*name="${route}"[^>]*/>`)
    return content.replace(matchRouteByName, '')
  }, routesContent)

  const routesWithoutEmptySet = layout
    ? removeEmtpySet(newRoutesContent, layout)
    : newRoutesContent

  writeFile(redwoodPaths.web.routes, routesWithoutEmptySet, {
    overwriteExisting: true,
  })
}

export const runCommandTask = async (commands, { verbose }) => {
  const tasks = new Listr(
    commands.map(({ title, cmd, args, opts = {} }) => ({
      title,
      task: async () => {
        return execa(cmd, args, {
          shell: true,
          cwd: `${getPaths().base}/api`,
          stdio: verbose ? 'inherit' : 'pipe',
          extendEnv: true,
          cleanup: true,
          ...opts,
        })
      },
    })),
    {
      renderer: verbose && VerboseRenderer,
      dateFormat: false,
    }
  )

  try {
    await tasks.run()
    return true
  } catch (e) {
    console.log(c.error(e.message))
    return false
  }
}

/*
 * Extract default CLI args from an exported builder
 */
export const getDefaultArgs = (builder) => {
  return Object.entries(builder).reduce(
    (options, [optionName, optionConfig]) => {
      // If a default is defined use it
      options[optionName] = optionConfig.default
      return options
    },
    {}
  )
}
