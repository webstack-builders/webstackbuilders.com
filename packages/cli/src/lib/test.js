/* eslint-env jest */

// Include at the top of your tests. Automatically mocks out the file system
//
// import { loadComponentFixture } from 'src/lib/test'
//
// test('true is true', () => {
//   expect('some output').toEqual(loadComponentFixture('component', 'filename.js'))
// })

import fs from 'fs'
import path from 'path'

jest.mock('@redwoodjs/internal', () => {
  const path = require('path')
  return {
    ...jest.requireActual('@redwoodjs/internal'),
    generate: () => {},
    getPaths: () => {
      const BASE_PATH = '/path/to/project'
      return {
        base: BASE_PATH,
        api: {
          dataMigrations: path.join(BASE_PATH, './api/prisma/dataMigrations'),
          db: path.join(global.__dirname, 'fixtures'), // this folder
          src: path.join(BASE_PATH, './api/src'),
          services: path.join(BASE_PATH, './api/src/services'),
          graphql: path.join(BASE_PATH, './api/src/graphql'),
          functions: path.join(BASE_PATH, './api/src/functions'),
        },
        web: {
          src: path.join(BASE_PATH, './web/src'),
          routes: path.join(BASE_PATH, 'web/src/Routes.js'),
          components: path.join(BASE_PATH, '/web/src/components'),
          layouts: path.join(BASE_PATH, '/web/src/layouts'),
          pages: path.join(BASE_PATH, '/web/src/pages'),
        },
        scripts: path.join(BASE_PATH, 'scripts'),
        generated: {
          base: path.join(BASE_PATH, '.redwood'),
          schema: path.join(BASE_PATH, '.redwood/schema.graphql'),
          types: {
            includes: path.join(BASE_PATH, '.redwood/types/includes'),
            mirror: path.join(BASE_PATH, '.redwood/types/mirror'),
          },
        },
      }
    },
  }
})

global.__prettierPath = path.resolve(
  __dirname,
  './__tests__/fixtures/prettier.config.js'
)

jest.mock('path', () => {
  const path = jest.requireActual('path')
  return {
    ...path,
    join: (...paths) => {
      if (
        paths &&
        paths[0] === '/path/to/project' &&
        paths[1] === 'prettier.config.js'
      ) {
        return global.__prettierPath
      }
      return path.join(...paths)
    },
  }
})

export const generatorsRootPath = path.join(
  __dirname,
  '..',
  'commands',
  'generate'
)

/**
 * Loads the fixture for a generator by assuming a lot of the path structure
 * automatically:
 *
 *   `loadGeneratorFixture('scaffold', 'NamePage.js')`
 *
 * will return the contents of:
 *
 *   `cli/src/commands/generate/scaffold/__tests__/fixtures/NamePage.js`
 */
export const loadGeneratorFixture = (generator, name) => {
  return loadFixture(
    path.join(
      __dirname,
      '..',
      'commands',
      'generate',
      generator,
      '__tests__',
      'fixtures',
      name
    )
  )
}

/**
 * Returns the contents of a text file in a `fixtures` directory
 */
export const loadFixture = (filepath) => {
  return fs.readFileSync(filepath).toString()
}
