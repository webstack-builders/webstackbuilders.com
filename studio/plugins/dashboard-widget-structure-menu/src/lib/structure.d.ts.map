{"version":3,"file":"structure.d.ts","sourceRoot":"","sources":["structure.ts"],"names":[],"mappings":"AAkBA,wBAAgB,cAAc,CAAC,KAAK,KAAA,OAKnC;AAED,wBAAgB,WAAW,CAAC,SAAS,KAAA,OASpC;AAED,wBAAgB,kBAAkB,CAAC,IAAI,KAAA,EAAE,OAAO,KAAA,EAAE,YAAY,UAAK,OAoBlE;AAED,wBAAgB,mBAAmB,QAOlC;AAKD,wBAAgB,aAAa,uCAmC5B"}