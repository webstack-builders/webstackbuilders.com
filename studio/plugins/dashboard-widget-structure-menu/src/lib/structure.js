"use strict";
/* global __DEV__ */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
exports.__esModule = true;
exports.loadStructure = exports.getDefaultStructure = exports.serializeStructure = exports.isStructure = exports.isSubscribable = void 0;
var structure_builder_1 = __importDefault(require("@sanity/desk-tool/structure-builder"));
var rxjs_1 = require("rxjs");
var operators_1 = require("rxjs/operators");
var prevStructureError = null;
if (__DEV__) {
    if (module.hot && module.hot.data) {
        prevStructureError = module.hot.data.prevError;
    }
}
function isSubscribable(thing) {
    return (thing &&
        (typeof thing.then === 'function' || typeof thing.subscribe === 'function'));
}
exports.isSubscribable = isSubscribable;
function isStructure(structure) {
    return (structure &&
        (typeof structure === 'function' ||
            typeof structure.serialize !== 'function' ||
            typeof structure.then !== 'function' ||
            typeof structure.subscribe !== 'function' ||
            typeof structure.type !== 'string'));
}
exports.isStructure = isStructure;
function serializeStructure(item, context, resolverArgs) {
    if (resolverArgs === void 0) { resolverArgs = []; }
    // Lazy
    if (typeof item === 'function') {
        return serializeStructure(item.apply(void 0, resolverArgs), context, resolverArgs);
    }
    // Promise/observable returning a function, builder or plain JSON structure
    if (isSubscribable(item)) {
        return rxjs_1.from(item).pipe(operators_1.mergeMap(function (val) { return serializeStructure(val, context, resolverArgs); }));
    }
    // Builder?
    if (item && typeof item.serialize === 'function') {
        return serializeStructure(item.serialize(context));
    }
    // Plain value?
    return rxjs_1.of(item);
}
exports.serializeStructure = serializeStructure;
function getDefaultStructure() {
    var items = structure_builder_1["default"].documentTypeListItems();
    return structure_builder_1["default"].list()
        .id('__root__')
        .title('Content')
        .showIcons(items.some(function (item) { return item.getSchemaType().icon; }))
        .items(items);
}
exports.getDefaultStructure = getDefaultStructure;
// We are lazy-requiring/resolving the structure inside of a function in order to catch errors
// on the root-level of the module. Any loading errors will be caught and emitted as errors
// eslint-disable-next-line complexity
function loadStructure() {
    var structure;
    try {
        var mod = require('part:@sanity/desk-tool/structure?') || getDefaultStructure();
        structure = mod && mod.__esModule ? mod["default"] : mod;
        // On invalid modules, when HMR kicks in, we sometimes get an empty object back when the
        // source has changed without fixing the problem. In this case, keep showing the error
        if (__DEV__ &&
            prevStructureError &&
            structure &&
            structure.constructor.name === 'Object' &&
            Object.keys(structure).length === 0) {
            return rxjs_1.throwError(prevStructureError);
        }
        prevStructureError = null;
    }
    catch (err) {
        prevStructureError = err;
        return rxjs_1.throwError(err);
    }
    if (!isStructure(structure)) {
        return rxjs_1.throwError(new Error("Structure needs to export a function, an observable, a promise or a structure builder, got " + typeof structure));
    }
    // Defer to catch immediately thrown errors on serialization
    return rxjs_1.defer(function () { return serializeStructure(structure); });
}
exports.loadStructure = loadStructure;
//# sourceMappingURL=structure.js.map