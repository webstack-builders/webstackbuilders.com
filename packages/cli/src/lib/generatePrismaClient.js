// helper used in Dev and Build commands

import fs from 'fs'
import path from 'path'

import { runCommandTask, getPaths } from '../lib'

const skipTask = (schema = getPaths().api.dbSchema) => {
  if (!fs.existsSync(schema)) {
    console.log(
      `Skipping database and Prisma client generation, no \`schema.prisma\` file found: \`${schema}\``
    )
    return true
  }
  return false
}

export const generatePrismaCommand = (schema) => {
  if (skipTask(schema)) {
    return {}
  }

  return {
    cmd: 'yarn prisma',
    args: ['generate', schema && `--schema="${schema}"`],
  }
}

/**
 * Conditionally generate the prisma client, skip if it already exists.
 */
export const generatePrismaClient = async ({
  verbose = true,
  force = true,
  schema = getPaths().api.dbSchema,
}) => {
  if (skipTask(schema)) {
    return
  }

  // Do not generate the Prisma client if it exists.
  if (!force) {
    // The Prisma client throws if it is not generated.
    try {
      // Import the client from the redwood apps node_modules path.
      const { PrismaClient } = require(path.join(
        getPaths().base,
        'node_modules/.prisma/client'
      ))
      // eslint-disable-next-line
      new PrismaClient()
      return // Client exists, so abort.
    } catch (e) {
      // Swallow your pain, and generate.
    }
  }

  return await runCommandTask(
    [
      {
        title: 'Generating the Prisma client...',
        ...generatePrismaCommand(schema),
      },
    ],
    {
      verbose,
    }
  )
}
