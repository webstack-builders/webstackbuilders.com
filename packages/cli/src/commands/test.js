import execa from 'execa'
import terminalLink from 'terminal-link'

import { ensurePosixPath } from '@redwoodjs/internal'
import { getProject } from '@redwoodjs/structure'

import { getPaths } from '../lib'

// https://github.com/facebook/create-react-app/blob/cbad256a4aacfc3084be7ccf91aad87899c63564/packages/react-scripts/scripts/test.js#L39
function isInGitRepository() {
  try {
    execa.commandSync('git rev-parse --is-inside-work-tree')
    return true
  } catch (e) {
    return false
  }
}

function isInMercurialRepository() {
  try {
    execa.commandSync('hg --cwd . root')
    return true
  } catch (e) {
    return false
  }
}

export const command = 'test [filter..]'
export const description = 'Run Jest tests. Defaults to watch mode'
export const builder = (yargs) => {
  yargs
    .strict(false) // so that we can forward arguments to jest
    .positional('filter', {
      default: getProject().sides,
      description:
        'Which side(s) to test, and/or a regular expression to match against your test files to filter by',
      type: 'array',
    })
    .option('watch', {
      describe:
        'Run tests related to changed files based on hg/git. Specify the name or path to a file to focus on a specific set of tests',
      type: 'boolean',
      default: true,
    })
    .option('collect-coverage', {
      describe:
        'Show test coverage summary and output info to coverage directory',
      type: 'boolean',
      default: false,
    })
    .option('db-push', {
      describe:
        "Syncs the test database with your Prisma schema without requiring a migration. It creates a test database if it doesn't already exist.",
      type: 'boolean',
      default: true,
    })
    .epilogue(
      `Also see the ${terminalLink(
        'Redwood CLI Reference',
        'https://redwoodjs.com/reference/command-line-interface#test'
      )}`
    )
}

export const handler = async ({
  filter: filterParams = [],
  watch = true,
  collectCoverage = false,
  dbPush = true,
  ...others
}) => {
  const rwjsPaths = getPaths()
  const forwardJestFlags = Object.keys(others).flatMap((flagName) => {
    if (
      ['watch', 'collect-coverage', 'db-push', '$0', '_'].includes(flagName)
    ) {
      // filter out flags meant for the rw test command only
      return []
    } else {
      // and forward on the other flags
      return [
        flagName.length > 1 ? `--${flagName}` : `-${flagName}`,
        others[flagName],
      ]
    }
  })

  // Only the side params
  const sides = filterParams.filter((filterString) =>
    getProject().sides.includes(filterString)
  )

  // All the other params, apart from sides
  const jestFilterArgs = [
    ...filterParams.filter(
      (filterString) => !getProject().sides.includes(filterString)
    ),
  ]

  const jestArgs = [
    ...jestFilterArgs,
    ...forwardJestFlags,
    collectCoverage ? '--collectCoverage' : null,
    '--passWithNoTests',
    ...jestFilterArgs,
    sides.includes('api') && '--runInBand',
  ].filter((flagOrValue) => flagOrValue !== null) // Filter out nulls, not booleans because user may have passed a --something false flag

  // If the user wants to watch, set the proper watch flag based on what kind of repo this is
  // because of https://github.com/facebook/create-react-app/issues/5210
  if (watch && !process.env.CI && !collectCoverage) {
    const hasSourceControl = isInGitRepository() || isInMercurialRepository()
    jestArgs.push(hasSourceControl ? '--watch' : '--watchAll')
  }

  // if no sides declared with yargs, default to all sides
  if (!sides.length) {
    getProject().sides.forEach((side) => sides.push(side))
  }

  jestArgs.push(
    '--config',
    `"${require.resolve('@redwoodjs/testing/config/jest/jest.config.js')}"`
  )

  if (sides.length > 0) {
    jestArgs.push('--projects', ...sides)
  }

  try {
    const cacheDirDb = `file:${ensurePosixPath(
      rwjsPaths.generated.base
    )}/test.db`
    const DATABASE_URL = process.env.TEST_DATABASE_URL || cacheDirDb

    if (sides.includes('api') && dbPush) {
      // Sync||create test database
      await execa(
        `yarn rw`,
        ['prisma db push', '--force-reset', '--accept-data-loss'],
        {
          cwd: rwjsPaths.api.base,
          stdio: 'inherit',
          shell: true,
          env: { DATABASE_URL },
        }
      )
    }

    // **NOTE** There is no official way to run Jest programmatically,
    // so we're running it via execa, since `jest.run()` is a bit unstable.
    // https://github.com/facebook/jest/issues/5048
    await execa('yarn jest', jestArgs, {
      cwd: rwjsPaths.base,
      shell: true,
      stdio: 'inherit',
      env: { DATABASE_URL },
    })
  } catch (e) {
    // Errors already shown from execa inherited stderr
    process.exit(e?.exitCode || 1)
  }
}
