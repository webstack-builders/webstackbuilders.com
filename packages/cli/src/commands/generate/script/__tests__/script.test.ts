global.__dirname = __dirname
// Load shared mocks
import '../../../../lib/test'

import path from 'path'

import { files } from '../script'

beforeAll(() => {})

test('creates a JavaScript function to execute', () => {
  const output = files({
    name: 'scriptyMcScript',
    typescript: false,
  })

  const expectedOutputPath = path.normalize(
    '/path/to/project/scripts/scriptyMcScript.js'
  )

  expect(Object.keys(output)).toContainEqual(expectedOutputPath)
  expect(output[expectedOutputPath]).toMatchSnapshot()
})

test('creates a TypeScript function to execute', () => {
  const output = files({
    name: 'typescriptyTypescript',
    typescript: true,
  })

  const expectedOutputPath = path.normalize(
    '/path/to/project/scripts/typescriptyTypescript.ts'
  )

  expect(Object.keys(output)).toContainEqual(expectedOutputPath)
  expect(output[expectedOutputPath]).toMatchSnapshot()
})
