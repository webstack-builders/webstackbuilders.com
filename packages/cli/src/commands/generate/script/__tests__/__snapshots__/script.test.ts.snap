// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates a JavaScript function to execute 1`] = `
"// To access your database
import { db } from '../api/src/lib/db'

export default async ({ args }) => {
  // Your script here...
  console.log(':: Executing script with args ::')
  console.log(args)
}
"
`;

exports[`creates a TypeScript function to execute 1`] = `
"// To access your database
import { db } from '../api/src/lib/db'

export default async ({ args }) => {
  // Your script here...
  console.log(':: Executing script with args ::')
  console.log(args)
}
"
`;
